File: module2/class1/exemplo26.cs
/*
================================================================================
ATIVIDADE PRÁTICA 26 - REFERENCE TYPE OVERHEAD (C#)
================================================================================

OBJETIVO:
- Demonstrar overhead de reference types vs value types
- Usar Memory/CPU profiler para identificar allocation overhead
- Otimizar usando value types e structs
- Medir diferença entre class vs struct performance

PROBLEMA:
- Reference types require heap allocation
- GC pressure from many small objects
- Memory profiler mostrará excessive allocations

SOLUÇÃO:
- Use value types para small data
- Struct instead of class quando apropriado
- Reduce boxing/unboxing overhead

================================================================================
*/

using System;
using System.Diagnostics;

// PERFORMANCE ISSUE: Reference type for small data
class PointClass {
    public double X { get; set; }
    public double Y { get; set; }
    
    public PointClass(double x, double y) {
        X = x;
        Y = y;
    }
    
    public double DistanceFromOrigin() {
        return Math.Sqrt(X * X + Y * Y);
    }
}

class Program {
    static void DemonstrateReferenceTypeOverhead() {
        Console.WriteLine("Starting reference type overhead demonstration...");
        Console.WriteLine("Monitor Memory profiler - should see excessive heap allocations");
        
        const int ITERATIONS = 1000000;
        
        long initialMemory = GC.GetTotalMemory(false);
        int initialCollections = GC.CollectionCount(0);
        
        var sw = Stopwatch.StartNew();
        
        double totalDistance = 0;
        
        for (int i = 0; i < ITERATIONS; i++) {
            // PERFORMANCE ISSUE: Heap allocation for each point
            var point = new PointClass(i * 0.1, i * 0.2); // Heap allocation
            totalDistance += point.DistanceFromOrigin();
            
            if (i % 100000 == 0) {
                Console.WriteLine($"Reference type: {i}/{ITERATIONS}, Memory: {GC.GetTotalMemory(false):N0} bytes");
            }
        }
        
        sw.Stop();
        
        long finalMemory = GC.GetTotalMemory(false);
        int finalCollections = GC.CollectionCount(0);
        
        Console.WriteLine($"Reference type overhead completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Total distance: {totalDistance:F2}");
        Console.WriteLine($"Memory growth: {(finalMemory - initialMemory):N0} bytes");
        Console.WriteLine($"GC collections: {finalCollections - initialCollections}");
        Console.WriteLine($"Objects allocated: {ITERATIONS}");
    }
    
    static void Main() {
        Console.WriteLine("Starting reference vs value type demonstration...");
        Console.WriteLine("Task: Creating many small objects using reference types");
        Console.WriteLine("Monitor Memory Usage Tool for allocation overhead");
        Console.WriteLine();
        
        DemonstrateReferenceTypeOverhead();
        
        Console.WriteLine();
        Console.WriteLine("=== PROFILING ANALYSIS ===");
        Console.WriteLine("Check Memory profiler for:");
        Console.WriteLine("- High number of heap allocations");
        Console.WriteLine("- GC pressure from many small objects");
        Console.WriteLine("- Memory overhead from reference types");
    }
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR VALUE TYPES)
================================================================================

using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;

// CORREÇÃO: Value type for small data
public readonly struct PointStruct {
    public double X { get; }
    public double Y { get; }
    
    public PointStruct(double x, double y) {
        X = x;
        Y = y;
    }
    
    public double DistanceFromOrigin() {
        return Math.Sqrt(X * X + Y * Y);
    }
}

// CORREÇÃO: ref struct for even better performance (stack-only)
public ref struct PointRefStruct {
    public double X;
    public double Y;
    
    public PointRefStruct(double x, double y) {
        X = x;
        Y = y;
    }
    
    public readonly double DistanceFromOrigin() {
        return Math.Sqrt(X * X + Y * Y);
    }
}

class Program {
    static void DemonstrateValueTypeOptimization() {
        Console.WriteLine("Starting value type optimization demonstration...");
        Console.WriteLine("Monitor Memory profiler - should see reduced allocations");
        
        const int ITERATIONS = 1000000;
        
        long initialMemory = GC.GetTotalMemory(false);
        int initialCollections = GC.CollectionCount(0);
        
        var sw = Stopwatch.StartNew();
        
        double totalDistance = 0;
        
        for (int i = 0; i < ITERATIONS; i++) {
            // CORREÇÃO: Stack allocation - no heap allocation
            var point = new PointStruct(i * 0.1, i * 0.2); // Stack allocation
            totalDistance += point.DistanceFromOrigin();
            
            if (i % 100000 == 0) {
                Console.WriteLine($"Value type: {i}/{ITERATIONS}, Memory: {GC.GetTotalMemory(false):N0} bytes");
            }
        }
        
        sw.Stop();
        
        long finalMemory = GC.GetTotalMemory(false);
        int finalCollections = GC.CollectionCount(0);
        
        Console.WriteLine($"Value type optimization completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Total distance: {totalDistance:F2}");
        Console.WriteLine($"Memory growth: {(finalMemory - initialMemory):N0} bytes");
        Console.WriteLine($"GC collections: {finalCollections - initialCollections}");
        Console.WriteLine("No heap allocations - all stack-based");
    }
    
    static void DemonstrateRefStructPerformance() {
        Console.WriteLine("Starting ref struct performance demonstration...");
        
        const int ITERATIONS = 1000000;
        
        var sw = Stopwatch.StartNew();
        
        double totalDistance = 0;
        
        for (int i = 0; i < ITERATIONS; i++) {
            // CORREÇÃO: ref struct - guaranteed stack allocation
            var point = new PointRefStruct(i * 0.1, i * 0.2);
            totalDistance += point.DistanceFromOrigin();
        }
        
        sw.Stop();
        
        Console.WriteLine($"Ref struct performance completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Total distance: {totalDistance:F2}");
        Console.WriteLine("Ref struct guarantees stack allocation - zero GC pressure");
    }
    
    static void DemonstrateSpanPerformance() {
        Console.WriteLine("Starting Span<T> performance demonstration...");
        
        const int ARRAY_SIZE = 1000000;
        
        // CORREÇÃO: Use Span<T> for efficient array slicing without allocation
        var data = new double[ARRAY_SIZE];
        for (int i = 0; i < ARRAY_SIZE; i++) {
            data[i] = i * 0.001;
        }
        
        var sw = Stopwatch.StartNew();
        
        double sum = 0;
        
        // CORREÇÃO: Span operations don't allocate
        Span<double> span = data.AsSpan();
        
        for (int i = 0; i < ARRAY_SIZE - 1; i++) {
            // Efficient slice operation - no allocation
            Span<double> slice = span.Slice(i, 2);
            sum += slice[0] * slice[1];
        }
        
        sw.Stop();
        
        Console.WriteLine($"Span<T> performance completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Sum: {sum:F2}");
        Console.WriteLine("Span<T> operations created no heap allocations");
    }
    
    // CORREÇÃO: Demonstrate avoiding boxing
    static void DemonstrateBoxingAvoidance() {
        Console.WriteLine("Starting boxing avoidance demonstration...");
        
        const int ITERATIONS = 1000000;
        
        var sw = Stopwatch.StartNew();
        
        // CORREÇÃO: Generic collection avoids boxing
        var points = new PointStruct[ITERATIONS];
        
        for (int i = 0; i < ITERATIONS; i++) {
            points[i] = new PointStruct(i * 0.1, i * 0.2);
        }
        
        double totalDistance = 0;
        foreach (var point in points) {
            totalDistance += point.DistanceFromOrigin(); // No boxing
        }
        
        sw.Stop();
        
        Console.WriteLine($"Boxing avoidance completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Total distance: {totalDistance:F2}");
        Console.WriteLine("Generic collections avoid boxing value types");
    }
    
    // CORREÇÃO: Demonstrate stackalloc for maximum performance
    static unsafe void DemonstrateStackAlloc() {
        Console.WriteLine("Starting stackalloc demonstration...");
        
        const int BUFFER_SIZE = 10000;
        
        var sw = Stopwatch.StartNew();
        
        // CORREÇÃO: Stack allocation for temporary buffers
        Span<double> buffer = stackalloc double[BUFFER_SIZE];
        
        // Fill buffer with calculations
        for (int i = 0; i < BUFFER_SIZE; i++) {
            buffer[i] = Math.Sin(i) * Math.Cos(i);
        }
        
        // Process buffer
        double sum = 0;
        foreach (var value in buffer) {
            sum += value;
        }
        
        sw.Stop();
        
        Console.WriteLine($"Stackalloc completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Sum: {sum:F6}");
        Console.WriteLine("Stackalloc provided zero-allocation temporary storage");
    }
    
    static void Main() {
        Console.WriteLine("Starting optimized value type demonstration...");
        Console.WriteLine("Task: Using value types to eliminate heap allocations");
        Console.WriteLine("Monitor Memory Usage Tool for reduced allocation pressure");
        Console.WriteLine();
        
        DemonstrateValueTypeOptimization();
        Console.WriteLine();
        DemonstrateRefStructPerformance();
        Console.WriteLine();
        DemonstrateSpanPerformance();
        Console.WriteLine();
        DemonstrateBoxingAvoidance();
        Console.WriteLine();
        DemonstrateStackAlloc();
        
        Console.WriteLine();
        Console.WriteLine("=== OPTIMIZATION RESULTS ===");
        Console.WriteLine("Improvements:");
        Console.WriteLine("- Value types eliminate heap allocations");
        Console.WriteLine("- Ref structs guarantee stack allocation");
        Console.WriteLine("- Span<T> provides zero-allocation array operations");
        Console.WriteLine("- Generic collections avoid boxing");
        Console.WriteLine("- Stackalloc eliminates temporary allocations");
        Console.WriteLine("- Dramatically reduced GC pressure and memory usage");
    }
}

================================================================================
*/

File: module2/class1/exemplo26.cpp
/*
================================================================================
ATIVIDADE PRÁTICA 26 - BRANCH PREDICTION FAILURES (C++)
================================================================================

OBJETIVO:
- Demonstrar impacto de branch misprediction na performance
- Usar CPU profiler para identificar branch prediction miss penalties
- Otimizar usando predictable branching patterns
- Medir diferença entre random vs predictable branches

PROBLEMA:
- Random branching causes pipeline stalls
- Branch misprediction penalties
- CPU Profiler mostrará branch miss statistics

SOLUÇÃO:
- Reorganizar código para predictable branches
- Use branchless programming quando apropriado
- Sort data para improve branch predictability

================================================================================
*/

#include <iostream>
#include <chrono>
#include <vector>
#include <random>
#include <algorithm>
using namespace std;

void demonstrateUnpredictableBranching() {
    cout << "Starting unpredictable branching demonstration..." << endl;
    cout << "Monitor CPU profiler - should see branch misprediction penalties" << endl;
    
    const int DATA_SIZE = 10000000;
    vector<int> data(DATA_SIZE);
    
    // Fill with random data
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dis(0, 255);
    
    for (int i = 0; i < DATA_SIZE; i++) {
        data[i] = dis(gen);
    }
    
    auto start = chrono::high_resolution_clock::now();
    
    long long sum = 0;
    int threshold = 128;
    
    // PERFORMANCE ISSUE: Random branching causes mispredictions
    for (int i = 0; i < DATA_SIZE; i++) {
        if (data[i] >= threshold) {  // Unpredictable branch - 50/50 chance
            sum += data[i];          // This branch is hard to predict
        } else {
            sum -= data[i];          // Random pattern causes pipeline stalls
        }
        
        if (i % 1000000 == 0) {
            cout << "Unpredictable branching: " << i << "/" << DATA_SIZE << endl;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Unpredictable branching completed in: " << duration.count() << " ms" << endl;
    cout << "Sum: " << sum << endl;
    cout << "Random branch pattern caused many mispredictions" << endl;
}

int main() {
    cout << "Starting branch prediction performance demonstration..." << endl;
    cout << "Task: Processing data with unpredictable branching patterns" << endl;
    cout << "Monitor CPU Usage Tool for branch misprediction overhead" << endl << endl;
    
    demonstrateUnpredictableBranching();
    
    cout << endl << "=== PROFILING ANALYSIS ===" << endl;
    cout << "Check CPU profiler for:" << endl;
    cout << "- Branch misprediction statistics" << endl;
    cout << "- Pipeline stall cycles" << endl;
    cout << "- Instructions per cycle degradation" << endl;
    
    return 0;
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR PREDICTABLE BRANCHES)
================================================================================

#include <iostream>
#include <chrono>
#include <vector>
#include <random>
#include <algorithm>
using namespace std;

void demonstratePredictableBranching() {
    cout << "Starting predictable branching demonstration..." << endl;
    cout << "Monitor CPU profiler - should see improved branch prediction" << endl;
    
    const int DATA_SIZE = 10000000;
    vector<int> data(DATA_SIZE);
    
    // Fill with random data
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dis(0, 255);
    
    for (int i = 0; i < DATA_SIZE; i++) {
        data[i] = dis(gen);
    }
    
    // CORREÇÃO: Sort data to make branches predictable
    sort(data.begin(), data.end());
    
    auto start = chrono::high_resolution_clock::now();
    
    long long sum = 0;
    int threshold = 128;
    
    // CORREÇÃO: Predictable branching pattern after sorting
    for (int i = 0; i < DATA_SIZE; i++) {
        if (data[i] >= threshold) {  // Predictable - all small values first, then large
            sum += data[i];          // Branch predictor learns pattern quickly
        } else {
            sum -= data[i];          // Very predictable due to sorted data
        }
        
        if (i % 1000000 == 0) {
            cout << "Predictable branching: " << i << "/" << DATA_SIZE << endl;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Predictable branching completed in: " << duration.count() << " ms" << endl;
    cout << "Sum: " << sum << endl;
    cout << "Sorted data enabled predictable branch patterns" << endl;
}

void demonstrateBranchlessCode() {
    cout << "Starting branchless code demonstration..." << endl;
    
    const int DATA_SIZE = 10000000;
    vector<int> data(DATA_SIZE);
    
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dis(0, 255);
    
    for (int i = 0; i < DATA_SIZE; i++) {
        data[i] = dis(gen);
    }
    
    auto start = chrono::high_resolution_clock::now();
    
    long long sum = 0;
    int threshold = 128;
    
    // CORREÇÃO: Branchless programming eliminates mispredictions
    for (int i = 0; i < DATA_SIZE; i++) {
        // Branchless conditional using arithmetic
        int condition = (data[i] >= threshold);  // 0 or 1
        sum += condition * data[i];              // Add if >= threshold
        sum -= (1 - condition) * data[i];       // Subtract if < threshold
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Branchless code completed in: " << duration.count() << " ms" << endl;
    cout << "Sum: " << sum << endl;
    cout << "No branches = no mispredictions" << endl;
}

void demonstrateOptimizedBranchLayout() {
    cout << "Starting optimized branch layout demonstration..." << endl;
    
    const int DATA_SIZE = 5000000;
    vector<int> data(DATA_SIZE);
    
    // Create data with known distribution for optimization
    for (int i = 0; i < DATA_SIZE; i++) {
        data[i] = i % 1000; // Predictable pattern
    }
    
    auto start = chrono::high_resolution_clock::now();
    
    long long sum = 0;
    
    // CORREÇÃO: Optimize branch layout for common case
    for (int i = 0; i < DATA_SIZE; i++) {
        // Most values will be < 800, so optimize for this path
        if (data[i] < 800) {  // Common case first - better branch prediction
            sum += data[i] * 2;
        } else if (data[i] < 900) {  // Less common case
            sum += data[i] * 3;
        } else {  // Rare case last
            sum += data[i] * 5;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Optimized branch layout completed in: " << duration.count() << " ms" << endl;
    cout << "Sum: " << sum << endl;
    cout << "Common-case-first layout improved prediction" << endl;
}

void demonstrateConditionalMove() {
    cout << "Starting conditional move demonstration..." << endl;
    
    const int DATA_SIZE = 10000000;
    vector<int> data(DATA_SIZE);
    
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dis(0, 1000);
    
    for (int i = 0; i < DATA_SIZE; i++) {
        data[i] = dis(gen);
    }
    
    auto start = chrono::high_resolution_clock::now();
    
    // CORREÇÃO: Using conditional move (cmov) instead of branches
    for (int i = 0; i < DATA_SIZE; i++) {
        int threshold = 500;
        int value1 = data[i] * 2;
        int value2 = data[i] / 2;
        
        // Modern compilers can optimize this to conditional move instruction
        data[i] = (data[i] >= threshold) ? value1 : value2;
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Conditional move completed in: " << duration.count() << " ms" << endl;
    cout << "Conditional moves avoid branch misprediction penalties" << endl;
}

int main() {
    cout << "Starting optimized branch prediction demonstration..." << endl;
    cout << "Task: Optimizing code for better branch prediction" << endl;
    cout << "Monitor CPU Usage Tool for improved branch prediction performance" << endl << endl;
    
    demonstratePredictableBranching();
    cout << endl;
    demonstrateBranchlessCode();
    cout << endl;
    demonstrateOptimizedBranchLayout();
    cout << endl;
    demonstrateConditionalMove();
    
    cout << endl << "=== OPTIMIZATION RESULTS ===" << endl;
    cout << "Improvements:" << endl;
    cout << "- Sorted data creates predictable branch patterns" << endl;
    cout << "- Branchless code eliminates mispredictions entirely" << endl;
    cout << "- Common-case-first layout improves prediction accuracy" << endl;
    cout << "- Conditional moves avoid branch penalties" << endl;
    cout << "- Significantly better instructions per cycle" << endl;
    
    return 0;
}

================================================================================
*/
